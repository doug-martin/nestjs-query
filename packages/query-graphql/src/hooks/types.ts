export enum HookTypes {
  BEFORE_CREATE_ONE = 'BeforeCreateOne',
  BEFORE_CREATE_MANY = 'BeforeCreateMany',
  BEFORE_UPDATE_ONE = 'BeforeUpdateOne',
  BEFORE_UPDATE_MANY = 'BeforeUpdateMany',
  BEFORE_DELETE_ONE = 'BeforeDeleteOne',
  BEFORE_DELETE_MANY = 'BeforeDeleteMany',
  BEFORE_QUERY_MANY = 'BeforeQueryMany',
  BEFORE_FIND_ONE = 'BeforeFindOne',
}
