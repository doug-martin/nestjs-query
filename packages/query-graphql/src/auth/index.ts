export * from './authorizer';
export * from './default-crud.authorizer';
export * from './tokens';
