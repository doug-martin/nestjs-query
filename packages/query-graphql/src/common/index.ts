export { DTONamesOpts, getDTONames, DTONames, getDTOIdTypeOrDefault } from './dto.utils';
export * from './external.utils';
export * from './resolver.utils';
export * from './object.utils';
