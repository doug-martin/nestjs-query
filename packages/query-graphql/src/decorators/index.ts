export {
  FilterableField,
  FilterableFieldOptions,
  FilterableFieldDescriptor,
  getFilterableFields,
} from './filterable-field.decorator';
export { ResolverMethodOpts } from './resolver-method.decorator';
export {
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  Relation,
  FilterableRelation,
  RelationDecoratorOpts,
  RelationTypeFunc,
  getRelations,
} from './relation.decorator';
export * from './resolver-mutation.decorator';
export * from './resolver-query.decorator';
export * from './resolver-field.decorator';
export { Reference, ReferenceDecoratorOpts, ReferenceTypeFunc } from './reference.decorator';
export { ResolverSubscription, SubscriptionResolverMethodOpts } from './resolver-subscription.decorator';
export { InjectPubSub } from './inject-pub-sub.decorator';
export * from './skip-if.decorator';
export * from './aggregate-query-param.decorator';
export * from './hook.decorator';
export * from './decorator.utils';
export * from './hook-args.decorator';
export * from './authorizer.decorator';
export * from './inject-authorizer.decorator';
export * from './inject-custom-authorizer.decorator';
export * from './key-set.decorator';
export * from './authorize-filter.decorator';
export * from './query-options.decorator';
export * from './id-field.decorator';
