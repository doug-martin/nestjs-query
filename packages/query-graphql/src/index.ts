export * from './types';
export {
  FilterableField,
  FilterableFieldOptions,
  ResolverMethodOpts,
  Relation,
  FilterableRelation,
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  RelationTypeFunc,
  RelationDecoratorOpts,
  Reference,
  ReferenceTypeFunc,
  ReferenceDecoratorOpts,
  InjectPubSub,
  BeforeCreateOne,
  BeforeCreateMany,
  BeforeUpdateOne,
  BeforeUpdateMany,
  BeforeDeleteOne,
  BeforeDeleteMany,
  BeforeQueryMany,
  BeforeFindOne,
  InjectAuthorizer,
  InjectCustomAuthorizer,
  Authorize,
  AuthorizerFilter,
  RelationAuthorizerFilter,
  HookArgs,
  MutationHookArgs,
  KeySet,
  QueryOptions,
  IDField,
  IDFieldOptions,
} from './decorators';
export * from './resolvers';
export * from './federation';
export { DTONamesOpts } from './common';
export { NestjsQueryGraphQLModule } from './module';
export { AutoResolverOpts } from './providers';
export { pubSubToken, GraphQLPubSub } from './subscription';
export { Authorizer, CustomAuthorizer, AuthorizerOptions, AuthorizationContext, OperationGroup } from './auth';
export {
  Hook,
  HookTypes,
  BeforeCreateOneHook,
  BeforeCreateManyHook,
  BeforeUpdateOneHook,
  BeforeUpdateManyHook,
  BeforeDeleteOneHook,
  BeforeDeleteManyHook,
  BeforeQueryManyHook,
  BeforeFindOneHook,
} from './hooks';
export { AuthorizerInterceptor, AuthorizerContext, HookInterceptor, HookContext } from './interceptors';
