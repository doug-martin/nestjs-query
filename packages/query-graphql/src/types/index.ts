export {
  PageInfoType,
  EdgeType,
  OffsetPageInfoType,
  ConnectionType,
  ConnectionOptions,
  CursorConnectionType,
  CursorConnectionOptions,
  ArrayConnectionType,
  ArrayConnectionOptions,
  OffsetConnectionType,
  OffsetConnectionOptions,
  StaticConnectionType,
  InferConnectionTypeFromStrategy,
} from './connection';
export { MutationArgsType } from './mutation-args.type';
export {
  FilterType,
  UpdateFilterType,
  DeleteFilterType,
  SubscriptionFilterType,
  QueryArgsType,
  QueryArgsTypeOpts,
  PagingTypes,
  CursorQueryArgsType,
  OffsetQueryArgsType,
  NonePagingQueryArgsType,
  CursorPagingType,
  OffsetPagingType,
  PagingStrategies,
} from './query';
export { CreateManyInputType } from './create-many-input.type';
export { CreateOneInputType } from './create-one-input.type';
export { UpdateManyResponseType } from './update-many-response.type';
export { UpdateOneInputType } from './update-one-input.type';
export { UpdateManyInputType } from './update-many-input.type';
export { DeleteManyResponseType } from './delete-many-reponse.type';
export { DeleteOneInputType } from './delete-one-input.type';
export { DeleteManyInputType } from './delete-many-input.type';
export { ConnectionCursorType, ConnectionCursorScalar } from './cursor.scalar';
export { RelationsInputType } from './relations-input.type';
export { RelationInputType } from './relation-input.type';
export { SubscriptionArgsType } from './subscription-args.type';
export { SubscriptionFilterInputType } from './subscription-filter-input.type';
export { AggregateResponseType, AggregateResponseOpts, AggregateArgsType } from './aggregate';
export { FindOneArgsType } from './find-one-args.type';
