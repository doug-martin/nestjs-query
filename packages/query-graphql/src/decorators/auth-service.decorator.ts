import { Class, MetaValue, ValueReflector } from '@nestjs-query/core';
import { CRUDAuthService } from '../auth';
import { AUTH_SERVICE_KEY } from './constants';

const reflector = new ValueReflector(AUTH_SERVICE_KEY);
export function AuthorizationService<DTO>(DTOClass: Class<DTO>) {
  return (AuthService: Class<CRUDAuthService<DTO>>) => {
    reflector.set(DTOClass, AuthService);
  };
}

export const getAuthService = <DTO>(DTOClass: Class<DTO>): MetaValue<Class<CRUDAuthService<DTO>>> => {
  return reflector.get(DTOClass);
};
