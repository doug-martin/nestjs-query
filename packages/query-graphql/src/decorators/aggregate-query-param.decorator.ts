import { AggregateQuery } from '@nestjs-query/core';
import { GraphQLResolveInfo } from 'graphql';
import { GqlExecutionContext } from '@nestjs/graphql';
import { createParamDecorator, ExecutionContext } from '@nestjs/common';
import graphqlFields from 'graphql-fields';

const EXCLUDED_FIELDS = ['__typename'];
const QUERY_OPERATORS: (keyof AggregateQuery<unknown>)[] = ['groupBy', 'count', 'avg', 'sum', 'min', 'max'];
export const AggregateQueryParam = createParamDecorator(<DTO>(data: unknown, ctx: ExecutionContext) => {
  const info = GqlExecutionContext.create(ctx).getInfo<GraphQLResolveInfo>();
  const fields = graphqlFields(info, {}, { excludedFields: EXCLUDED_FIELDS }) as Record<
    keyof AggregateQuery<DTO>,
    Record<keyof DTO, unknown>
  >;
  return QUERY_OPERATORS.filter((operator) => !!fields[operator]).reduce((query, operator) => {
    const queryFields = Object.keys(fields[operator]) as (keyof DTO)[];
    if (queryFields && queryFields.length) {
      return { ...query, [operator]: queryFields };
    }
    return query;
  }, {} as AggregateQuery<DTO>);
});
