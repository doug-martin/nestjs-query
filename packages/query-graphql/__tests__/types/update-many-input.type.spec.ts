import { plainToClass } from 'class-transformer';
import { MinLength, validateSync } from 'class-validator';
import { ObjectType, Resolver, Args, Query, InputType, Int } from '@nestjs/graphql';
import { FilterableField } from '../../src/decorators';
import { UpdateManyInputType } from '../../src/types';
import { generateSchema } from '../__fixtures__';

describe('UpdateManyInputType', (): void => {
  @InputType('FakeUpdateManyInput')
  @ObjectType()
  class FakeUpdateManyType {
    @FilterableField()
    @MinLength(5)
    name!: string;
  }
  @InputType()
  class UpdateMany extends UpdateManyInputType(FakeUpdateManyType, FakeUpdateManyType) {}

  it('should create an args type with the field as the type', async () => {
    @Resolver()
    class UpdateManyInputTypeSpec {
      @Query(() => Int)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      updateTest(@Args('input') input: UpdateMany): number {
        return 1;
      }
    }
    const schema = await generateSchema([UpdateManyInputTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('should return the input when accessing the update field', () => {
    const Type = UpdateManyInputType(FakeUpdateManyType, FakeUpdateManyType);
    const input = {};
    const it = plainToClass(Type, input);
    expect(it).toEqual(input);
  });

  describe('validation', () => {
    it('should validate the filter is not empty', () => {
      const input = { update: { name: 'hello world' } };
      const it = plainToClass(UpdateMany, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [],
          constraints: {
            isNotEmptyObject: 'filter must be a non-empty object',
          },
          property: 'filter',
          target: input,
        },
      ]);
    });

    it('should validate the update input', () => {
      const input = { filter: { name: { eq: 'hello world' } }, update: {} };
      const it = plainToClass(UpdateMany, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [
            {
              children: [],
              constraints: {
                minLength: 'name must be longer than or equal to 5 characters',
              },
              property: 'name',
              target: {},
            },
          ],
          property: 'update',
          target: input,
          value: input.update,
        },
      ]);
    });
  });
});
