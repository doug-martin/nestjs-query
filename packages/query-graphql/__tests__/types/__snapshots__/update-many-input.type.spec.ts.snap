// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateManyInputType should create an args type with the field as the type 1`] = `
type Query {
  updateTest(input: UpdateMany!): Int!
}

input UpdateMany {
  """Filter used to find fields to update"""
  filter: FakeUpdateManyTypeUpdateFilter!

  """The update to apply to all records found using the filter"""
  update: FakeUpdateManyInput!
}

input FakeUpdateManyTypeUpdateFilter {
  and: [FakeUpdateManyTypeUpdateFilter!]
  or: [FakeUpdateManyTypeUpdateFilter!]
  name: StringFieldComparison
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

input FakeUpdateManyInput {
  name: String!
}

`;
